runtime/parachains/src/dmp/migration.rs

```diff
@@ -0,0 +1,196 @@
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A module that is responsible for migration of DMP storage.
+
+use frame_support::{
+	ensure,
+	pallet_prelude::ValueQuery,
+	storage::migration::clear_storage_prefix,
+	storage_alias,
+	traits::{Get, PalletInfoAccess, StorageVersion},
+	Twox64Concat,
+};
+
+use primitives::v2::BlockNumber;
+use sp_std::vec::Vec;
+
+/// The current storage version.
+pub const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+pub mod v0 {
+	use super::*;
+
+	pub use primitives::v2::{Id as ParaId, InboundDownwardMessage};
+
+	#[storage_alias]
+	pub(crate) type DownwardMessageQueues<T: crate::dmp::Config> = StorageMap<
+		crate::dmp::Pallet<T>,
+		Twox64Concat,
+		ParaId,
+		Vec<InboundDownwardMessage<BlockNumber>>,
+		ValueQuery,
+	>;
+}
+pub mod v1 {
+	use super::*;
+	use crate::{configuration, dmp, dmp::Config};
+
+	pub fn pre_migrate<T: Config>() -> Result<(), &'static str> {
+		let count = v0::DownwardMessageQueues::<T>::iter()
+			.filter(|(para_id, messages)| {
+				if messages.len() > 0 {
+					log::debug!(
+						target: "runtime",
+						"dmp queue for para {} needs migration; message count = {}",
+						u32::from(*para_id),
+						messages.len()
+					);
+					true
+				} else {
+					false
+				}
+			})
+			.count();
+
+		log::info!(
+			target: "runtime",
+			"Migrating dmp storage for {} parachains",
+			count,
+		);
+
+		Ok(())
+	}
+
+	/// This migration converts the storage to a new representation which enables pagination.
+	pub fn migrate<T: Config>() -> frame_support::weights::Weight {
+		let config = <configuration::Pallet<T>>::config();
+		let version = StorageVersion::get::<dmp::Pallet<T>>();
+
+		log::info!(
+			target: "runtime",
+			"Version {:?} -> Version {:?}",
+			version, STORAGE_VERSION
+		);
+
+		if version < STORAGE_VERSION {
+			let weight = v0::DownwardMessageQueues::<T>::iter()
+				.map(|(para_id, messages)| {
+					let mut weight = 0u64;
+					for message in messages {
+						let queue_weight =
+							<dmp::Pallet<T>>::queue_downward_message(&config, para_id, message.msg)
+								.expect("Failed to queue a message");
+						weight = weight.saturating_add(
+							queue_weight.saturating_add(T::DbWeight::get().reads(1)),
+						);
+					}
+					weight
+				})
+				.sum::<u64>();
+
+			// Update current storage version to latest.
+			STORAGE_VERSION.put::<dmp::Pallet<T>>();
+
+			// Remove old queues, but we'll keep the MQC head around until collators receive their node/runtime upgrades.
+			let results = clear_storage_prefix(
+				<dmp::Pallet<T>>::name().as_bytes(),
+				b"DownwardMessageQueues",
+				&[],
+				None,
+				None,
+			);
+
+			weight.saturating_add(T::DbWeight::get().writes(results.unique as u64))
+		} else {
+			0
+		}
+	}
+
+	pub fn post_migrate<T: Config>() -> Result<(), &'static str> {
+		let old_storage_clean =
+			v0::DownwardMessageQueues::<T>::iter().any(|(_, messages)| messages.len() > 0);
```

Comment:
the name should be old_storage_not_empty? or change the calculation and ensure below